lib/ClangImporter/ImportType.cpp

```diff
@@ -2230,6 +2230,161 @@ ImportedType ClangImporter::Implementation::importFunctionParamsAndReturnType(
+  ImportTypeKind importKind = ImportTypeKind::Parameter;
+  if (param->hasAttr<clang::CFReturnsRetainedAttr>())
+    importKind = ImportTypeKind::CFRetainedOutParameter;
+  else if (param->hasAttr<clang::CFReturnsNotRetainedAttr>())
+    importKind = ImportTypeKind::CFUnretainedOutParameter;
+
+  return importKind;
+}
+
+Optional<ClangImporter::Implementation::ImportParameterTypeResult>
+ClangImporter::Implementation::importParameterType(
+    const clang::ParmVarDecl *param, OptionalTypeKind optionalityOfParam,
+    bool allowNSUIntegerAsInt, bool isNSDictionarySubscriptGetter,
+    bool paramIsError, bool paramIsCompletionHandler,
+    Optional<unsigned> completionHandlerErrorParamIndex,
+    ArrayRef<GenericTypeParamDecl *> genericParams,
+    llvm::function_ref<void(Diagnostic &&)> addImportDiagnosticFn) {
+  auto paramTy = param->getType();
+
+  ImportTypeKind importKind = getImportTypeKindForParam(param);
+
+  // Import the parameter type into Swift.
+  auto attrs = getImportTypeAttrs(param, /*isParam=*/true);
+  Type swiftParamTy;
+  bool isInOut = false;
+  bool isParamTypeImplicitlyUnwrapped = false;
+
+  // Sometimes we import unavailable typedefs as enums. If that's the case,
+  // use the enum, not the typedef here.
+  if (auto typedefType = dyn_cast<clang::TypedefType>(paramTy.getTypePtr())) {
+    if (isUnavailableInSwift(typedefType->getDecl())) {
+      if (auto clangEnum =
+              findAnonymousEnumForTypedef(SwiftContext, typedefType)) {
+        // If this fails, it means that we need a stronger predicate for
+        // determining the relationship between an enum and typedef.
+        assert(clangEnum.getValue()
+                   ->getIntegerType()
+                   ->getCanonicalTypeInternal() ==
+               typedefType->getCanonicalTypeInternal());
+        if (auto swiftEnum = importDecl(*clangEnum, CurrentVersion)) {
+          swiftParamTy = cast<NominalTypeDecl>(swiftEnum)->getDeclaredType();
+        }
+      }
+    }
+  } else if (isa<clang::PointerType>(paramTy) &&
+             isa<clang::TemplateTypeParmType>(paramTy->getPointeeType())) {
+    auto pointeeType = paramTy->getPointeeType();
+    auto templateParamType = cast<clang::TemplateTypeParmType>(pointeeType);
+    PointerTypeKind pointerKind = pointeeType.getQualifiers().hasConst()
+                                      ? PTK_UnsafePointer
+                                      : PTK_UnsafeMutablePointer;
+    auto genericType = findGenericTypeInGenericDecls(
+        *this, templateParamType, genericParams, attrs, addImportDiagnosticFn);
+    swiftParamTy = genericType->wrapInPointer(pointerKind);
+    if (!swiftParamTy)
+      return None;
+  } else if (isa<clang::ReferenceType>(paramTy) &&
+             isa<clang::TemplateTypeParmType>(paramTy->getPointeeType())) {
+    auto templateParamType =
+        cast<clang::TemplateTypeParmType>(paramTy->getPointeeType());
+    swiftParamTy = findGenericTypeInGenericDecls(
```

Comment:
It seems like this patch enables a lot of new behavior for Objective-C++ methods (for example, templated methods, if those are even allowed). Can you add some more tests? Otherwise people will use them and we'll have no idea if they work or not.